external/fv3fit/fv3fit/pytorch/graph/mpg_unet.py

```diff
@@ -0,0 +1,348 @@
+            out_feats=node_hidden_channels,
+            edge_func=edge_network,
+            aggregator_type=aggregator,
+        )
+        self.gru = nn.GRU(node_hidden_channels, node_hidden_channels)
+        self.relu = activation
+        self.out_feats = node_hidden_channels
+
+    def reset_parameters(self):
+        """Reinitialize model parameters."""
+        self.project_node_feats[0].reset_parameters()
+        self.gnn_layer.reset_parameters()
+        for layer in self.gnn_layer.edge_func:
+            if isinstance(layer, nn.Linear):
+                layer.reset_parameters()
+        self.gru.reset_parameters()
+
+    def forward(self, in_node_feats):
+        """Performs message passing and updates node representations.
+
+        Args:
+        g : DGL graph
+        in_node_feats : Input node features.
+
+        Returns:
+        out_node_feats : Output node representations.
+        """
+
+        out_node_feats = torch.zeros(
+            in_node_feats.size(0),
+            in_node_feats.size(1),
+            in_node_feats.size(2),
+            in_node_feats.size(3),
+            self.out_feats,
+        )  # initialize the updated node features (n_batch,tile,x,y,features)
+
+        for batch in range(
+            in_node_feats.size(0)
+        ):  # for loop over the n_batch since dgl NNConv
+            # just accpets data in (grids, features) format
+            node_feats = in_node_feats[batch].squeeze()
+            in_size = node_feats.size()
+            node_feats = node_feats.reshape(
+                node_feats.shape[0] * node_feats.shape[1] * node_feats.shape[2],
+                node_feats.shape[3],
+            )  # reashape (tile,x,y,features) to (tile*x*y,features)
+
+            node_feats = self.project_node_feats(
+                node_feats
+            )  # (grids , node_out_channels)
+            hidden_feats = node_feats.unsqueeze(0)  # (1, grids , node_out_channels)
+
+            for _ in range(self.num_step_message_passing):
+                node_feats = self.relu(
+                    self.gnn_layer(self.graph, node_feats, self.edge_relation)
+                )
+                node_feats, hidden_feats = self.gru(
+                    node_feats.unsqueeze(0), hidden_feats
+                )
+                node_feats = node_feats.squeeze(0)
+            out_node_feats[batch] = node_feats.reshape(
+                in_size[0], in_size[1], in_size[2], node_feats.size(1)
+            )  # reashape (tile*x*y,features) to (tile,x,y,features)
+        return out_node_feats
+
+
+class Down(nn.Module):
+    """
+    A class for the down path of the U-net which
+    reduce the resolution by applying pooling layer
+    """
+
+    def __init__(self, config):
+        super(Down, self).__init__()
+        self.pool = nn.AvgPool2d(
+            kernel_size=config.pooling_size, stride=config.pooling_stride
+        )
+
+    def forward(self, x):
+        input_size = x.size()
+        x = x.permute(
```

Comment:
Very optional: Can we somehow avoid permuting before and after every down and up operation by doing the permutation at the start and end of the model instead? I'm reluctant to ask because I know how annoying debugging shapes can be if you tried to switch this.

Reply:
You don't have to do this comment, but please reply telling me whether you plan to.

Reply:
I thought it would be nicer to have the U-Net code cleaner than Up or Down. That's the reason that I have those codes here.